src/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

```diff
@@ -93,31 +125,42 @@ public Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library, NuGe
                 // Extract the resolved node
                 tasks.Remove(task);
                 var dependencyNode = await task;
-
-                // Wire the node into the tree
                 dependencyNode.OuterNode = node;
+
                 node.InnerNodes.Add(dependencyNode);
             }
 
             return node;
         }
 
-        private Func<string, bool> ChainPredicate(Func<string, bool> predicate, GraphItem<RemoteResolveResult> item, LibraryDependency dependency)
+        private Func<LibraryRange, DependencyResult> ChainPredicate(Func<LibraryRange, DependencyResult> predicate, GraphNode<RemoteResolveResult> node, LibraryDependency dependency)
         {
-            return name =>
+            var item = node.Item;
+
+            return library =>
+            {
+                if (item.Data.Match.Library.Name == library.Name)
                 {
-                    if (item.Data.Match.Library.Name == name)
-                    {
-                        throw new Exception(string.Format("Circular dependency references not supported. Package '{0}'.", name));
-                    }
+                    return DependencyResult.Cycle;
+                }
 
-                    if (item.Data.Dependencies.Any(d => d != dependency && d.Name == name))
+                foreach (var d in item.Data.Dependencies)
+                {
+                    if (d != dependency && d.Name == library.Name)
                     {
-                        return false;
+                        if (d.LibraryRange.VersionRange != null &&
+                            library.VersionRange != null &&
+                            d.LibraryRange.VersionRange.MinVersion < library.VersionRange.MinVersion)
+                        {
+                            return DependencyResult.PotentiallyDowngraded;
```

Comment:
Just so I'm clear, if a `PotentiallyDowngraded` dependency is not actually downgraded, it's `Eclipsed`, right? Just confirming that we don't need to walk down the dependencies of a `PotentiallyDowngraded` node because no matter what, it will never be used in the final graph.